### PR TITLE
prefetch-yarn-deps: Fix handling of scoped packages

### PR DESCRIPTION
**Pull request by lorenzleutgeb**
## Description of changes

I ran into an issue with a lockfile entry of the form `"@org/somepack@file:vendor/orgpacks/somepack/assets"`. `fetch-yarn-deps` would crash, because it would not detect `@file:...` and just blindly access `pkg.resolved`, which was not present.

## Things done

Diagnosed and fixed the bug, added a test to avoid regressions in the future.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lorenzleutgeb** [comment]
@ofborg build gitmoji-cli marp

File: pkgs/build-support/node/fetch-yarn-deps/tests/file.lock [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
+"@org/somepack@file:vendor/orgpacks/somepack/assets":
```
Comment (yu-re-ka) [inline pkgs/build-support/node/fetch-yarn-deps/tests/file.lock:5]: Please help me to understand the example. In the log of this test, I see only one line `ignoring lockfile entry "otherpack" which points at path "vendor/otherpack"`. Why does the `somepack` not enter the `ignoring lockfile ...` code path?

Comment (lorenzleutgeb) [reply pkgs/build-support/node/fetch-yarn-deps/tests/file.lock:5]: You're right. There was another bug in https://github.com/NixOS/nixpkgs/blob/11f4e814f624face2e8c775a8bd892be6b8b811f/pkgs/build-support/node/fetch-yarn-deps/index.js#L140 which accesses `pkg.resolved`, which will fail for `file:...` entries. I took a closer look and simplified that function...

**Comment by lorenzleutgeb** [comment]
@ofborg build gitmoji-cli marp

**Comment by lorenzleutgeb** [comment]
@lilyinstarlight maybe this is worth your attention before 23.11.

**Comment by lilyinstarlight** [comment]
@ofborg eval
@ofborg build prefetch-yarn-deps prefetch-yarn-deps.tests

Requesting a fresh eval and also fetcher test build, since this eval is a bit stale and because the commit message uses `fetch-yarn-deps` instead of `prefetch-yarn-deps`, ofborg doesn't build the tests

**Review by lilyinstarlight — APPROVED** [review]
The diff itself looks fine to me and does not appear to be a breaking change, so if ofborg builds pass and @yu-re-ka has no objections, I think this looks fine to merge

Thank you!

**Comment by lilyinstarlight** [comment]
Ofborg failed because fetcher should probably set `SSL_CERT_FILE` like the others, now that #257513 was merged and Node.js no longer uses the built-in ones... (and this probably means _all_ yarn fetchers are failing/unreproducible FODs right now...)

I'll open a separate PR to fix that since it is orthogonal to this PR

**Comment by lilyinstarlight** [comment]
I've opened #269061 to fix the fetcher failures

**Pull request merged**